src/dems-engine/dems-engine.service.ts

```diff
@@ -0,0 +1,542 @@
+    return {
+      transaction: payload,
+      TxTp: transactionType,
+      dataCache,
+    };
+  }
+
+  /**
+   * Saves transaction data and sends notification to event director
+   * @param tazamaPayload The Tazama payload to process
+   * @param transactionType The transaction type
+   * @param endToEndId The end-to-end ID for the transaction
+   * @param TransactionDetails The transaction relationship data
+   */
+  @ApmSpan('dems-save-transaction-and-notify')
+  async saveTransactionDataAndNotify(tazamaPayload: TazamaPayload, transactionType: string, endToEndId: string): Promise<void> {
+    try {
+      await Promise.all([
+        this.databaseOperationsService.saveTransactionHistory(tazamaPayload, `${transactionType}_${endToEndId}`),
+        // this.databaseOperationsService.saveTransactionRelationship(transactionRelationship),
+        this.natsService.notifyEventDirector(tazamaPayload),
+      ]);
+
+      this.loggerService.log('Successfully saved transaction history, transaction relationship, and notified event-director');
+    } catch (error) {
+      // Determine which operation failed for better error reporting
+      let errorMessage = 'Failed to complete transaction operations';
+
+      if (String(error).includes('saveTransactionHistory') || String(error).includes('transaction history')) {
+        errorMessage = `Failed to save transaction history: ${String(error)}`;
+      } else if (String(error).includes('saveTransactionRelationship') || String(error).includes('transaction relationship')) {
+        errorMessage = `Failed to save transaction relationship: ${String(error)}`;
+      } else if (String(error).includes('notifyEventDirector') || String(error).includes('event-director')) {
+        errorMessage = `Failed to notify event-director: ${String(error)}`;
+      } else {
+        errorMessage = `Failed to complete transaction operations: ${String(error)}`;
+      }
+
+      this.loggerService.error(errorMessage, '', this.LOG_CONTEXT);
+      throw error;
+    }
+  }
+
+  /**
+   * Builds an error response object
+   * @param message The error message
+   * @param differences Array of validation differences
+   * @param schema Optional schema object
+   * @returns Formatted error response
+   */
+  private buildErrorResponse(message: string, differences: string[], schema?: any): ErrorResponse {
+    this.loggerService.warn(`Building error response: ${message}`, this.LOG_CONTEXT);
+    return {
+      isMatch: false,
+      message,
+      differences,
+      ...(schema && { schema }),
+    };
+  }
+
+  @ApmSpan('dems-handle-message')
+  async handleMessage(payload: any, endpoint: string, tenantId: string, isPayloadXml: boolean): Promise<ErrorResponse | ProcessingResult> {
+    let transformedPayload: any; //contains the XML --> JSON converted payload
+
+    if (isPayloadXml) {
+      const schemaResult = await this.findSchemaAndMapping(endpoint);
```

Comment:
Im wondering if this needs to be in this `if` branch. We do the same operation here outside anyways: https://github.com/tazama-lf/event-monitoring-service/blob/e1144df6144c7616700c486d70e3d6b1b4fe9a5a/src/dems-engine/dems-engine.service.ts#L478